### PR TITLE
Add Explicit Sync/Async clients, and add mypy types for the library

### LINKED ISSUE #140: Add type hint support
The authzed library already ships with most of the necessary type hints in `*.pyi` files. However, mypy can't use this because:
1. There's no `py.typed` file in the package so these types can't be used at all
2. Even if the file were there, there's no types for `__init__.py`
3. In `__init__.py` the type for the Client is wrong/ really hard to implement because it could _either_ be the Async* versions or the Sync versions

E.g. depending on whether you're in an async loop, the correct definition is either
```python
class Client(SchemaServiceStub, PermissionsServiceStub, ExperimentalServiceStub, WatchServiceStub):
  ...
```

OR
```python
class Client(SchemaServiceAsyncStub, PermissionsServiceAsyncStub, ExperimentalServiceAsyncStub, WatchServiceAsyncStub):
  ...
```

So, largely we need to solve the `__init__.py` file before we can continue. I have one concrete suggestion: Make a separate `AsyncClient` which has the latter type signature. Then, make a new `SyncClient` which is just an alias to the existing Client. That will allow the type hints to be straightforward.

Then, if we add the type hints directly to the `__init__.py` file for the init method, we wouldn't need a separate `__init__.pyi` file since the init file would have all of the necessary type hints

### PR DESCRIPTION
**Pull request by intentionally-left-nil**
Fixes #140 
This adds mypy type support to the authzed library.
The code was almost there already, just needed types for the `__init__.py`

The challenge with `__init__.py` is that the Client is either sync or async depending on whether it's in an event loop. Typing this isn't really possible. So, per discord conversation, I broke out two new classes which explicitly are sync or async (and typed those appropriately)

Also, I added a new CI job to validate the types in the future

Test plan:
Made sure the py.typed file is present:
❯ tar -tf dist/authzed-0.14.0.tar.gz | grep py.typed
authzed-0.14.0/authzed/py.typed
❯ unzip -l dist/authzed-0.14.0-py3-none-any.whl | grep py.typed
        0  01-01-1980 00:00   authzed/py.typed

Made sure that mypy passes
poetry shell
mypy authzed

Made an invalid change to the __init__.pyi file and validated there was a type error
Created a CI job to ensure that the types pass as expected.

recheck

### CODE REVIEW

**Comment by intentionally-left-nil** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by intentionally-left-nil** [comment]
Couple notes:
1. The types have to be in a `__init__.pyi` file instead of the code file because the Async* definitions (e.g. SchemaServiceAsyncStub) don't exist in the *.py files, but only in the .pyi files. (Maybe there's some IF TYPE_CHECKING magic you could use but I thought this way was straightforward)
2. The `Client` doesn't have any type hints because of the root problem of typing something that is either sync or async
3. It's such a small client that I didn't bother trying to DRY anything up, but let me know if you would prefer for there to be changes

File: pyproject.toml [hunk 1]
```diff
@@ -31,11 +31,19 @@ pyflakes = "^3.0.1"
...
 [tool.black]
 exclude = '(^\.(.*)|^(.*)_pb2(_grpc)?\.py)'
 line-length = 100
 
+[tool.mypy]
+explicit_package_bases = true
+
+[[tool.mypy.overrides]]
+ignore_missing_imports = true
+module = ["google.rpc.*", "grpcutil"]
```
Comment (intentionally-left-nil) [inline pyproject.toml:45]: Types for these modules don't seem to exist ... anywhere so I just told mypy to ignore them

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -74,6 +74,6 @@ jobs:
           whereis protoc-gen-mypy
           whereis protoc-gen-mypy_grpc
           echo $GITHUB_PATH
-        uses: "bufbuild/buf-setup-action@v1.30.1"
+      - uses: "bufbuild/buf-setup-action@v1.30.1"
```
Comment (intentionally-left-nil) [inline .github/workflows/test.yaml:77]: This was previously broken and the test file wasn't running at all 😱 Fixing it here so it runs

**Review by josephschorr — APPROVED** [review]
LGTM

**Pull request merged**